Adopt BrowserEngineCore

justinmichaud opened the pull request:
#### 676f1380f1d50dd92f84b952b1fb8305a0a83224
<pre>
Adopt BrowserEngineCore
<a href="https://bugs.webkit.org/show_bug.cgi?id=268175">https://bugs.webkit.org/show_bug.cgi?id=268175</a>
<a href="https://rdar.apple.com/121675394">rdar://121675394</a>

Reviewed by Wenson Hsieh and Elliott Williams.

Adopt BrowserEngineCore, and compile-time assert that it is enabled.

* Source/JavaScriptCore/Configurations/Base.xcconfig:
* Source/JavaScriptCore/assembler/FastJITPermissions.h:
(threadSelfRestrictRWXToRW):
(threadSelfRestrictRWXToRX):
* Source/JavaScriptCore/jit/ExecutableAllocator.cpp:
(JSC::initializeJITPageReservation):
* Source/WTF/wtf/PlatformUse.h:
* Source/WebCore/Configurations/WebCore.xcconfig:

Canonical link: <a href="https://commits.webkit.org/273590@main">https://commits.webkit.org/273590@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fb5ab35a603e9407f14a6d11c76840b945b408b3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/35892 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/14835 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/38130 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/38614 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32292 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/37111 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/17256 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/11861 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/31047 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/36446 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/12547 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/31923 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11015 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/11035 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/32116 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/39864 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/30347 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/32616 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/32437 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/36975 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/35741 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11199 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9103 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35060 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12949 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/42445 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/11709 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/8803 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4660 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12029 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WTF/wtf/PlatformUse.h:
```diff
@@ -416,8 +416,10 @@
+#if (PLATFORM(IOS) || PLATFORM(MAC)) && !PLATFORM(IOS_FAMILY_SIMULATOR) && __has_include(<BrowserEngineCore/BEMemory.h>)
 #define USE_INLINE_JIT_PERMISSIONS_API 1
+#elif (PLATFORM(IOS) && __IPHONE_OS_VERSION_MAX_ALLOWED >= 170400  && __IPHONE_OS_VERSION_MAX_ALLOWED < 180000)
+#error
```
Maybe a word or two here would be nice?

emw-apple replied:
Can we just use SDK version:
```c
#if PLATFORM(IOS) && !PLATFORM(IOS_FAMILY_SIMULATOR) && \
__IPHONE_OS_VERSION_MAX_ALLOWED >= 170400 && \
__IPHONE_OS_VERSION_MAX_ALLOWED < 180000
```
as the condition? Testing for the header seems likely to cause this to suddenly flip on other platforms, and the error statement implies to me that we really want to be sure it's enabled in `17.4..<18`

whsieh commented on Source/WTF/wtf/PlatformUse.h:
```diff
@@ -416,8 +416,10 @@
 #endif
 
-#if PLATFORM(IOS) && !PLATFORM(IOS_FAMILY_SIMULATOR) && __has_include(<ServiceExtensionsCore/SEMemory_Private.h>)
+#if (PLATFORM(IOS) || PLATFORM(MAC)) && !PLATFORM(IOS_FAMILY_SIMULATOR) && __has_include(<BrowserEngineCore/BEMemory.h>)
```
(From chatting on Slack, it seems adding `PLATFORM(MAC)` wasn't intentional)

emw-apple commented on Source/JavaScriptCore/Configurations/Base.xcconfig:
```diff
@@ -160,7 +160,7 @@ WK_USE_RESTRICTED_ENTITLEMENTS = $(USE_INTERNAL_SDK);
 
 // Tomorrow me will be smart enough to figure out how to do this properly.
-JSC_SEC_LD_FLAGS[sdk=iphoneos*] = -weak_framework ServiceExtensionsCore;
+JSC_SEC_LD_FLAGS[sdk=iphoneos*] = -weak_framework BrowserEngineCore;
```
Discussed elsewhere, but I believe we need an additional `JSC_SEC_LD_FLAGS[sdk=iphoneos18.0*] = ;` clause for now.

Commit:
Adopt BrowserEngineCore
https://bugs.webkit.org/show_bug.cgi?id=268175
rdar://121675394

Reviewed by Wenson Hsieh and Elliott Williams.

Adopt BrowserEngineCore, and compile-time assert that it is enabled.

* Source/JavaScriptCore/Configurations/Base.xcconfig:
* Source/JavaScriptCore/assembler/FastJITPermissions.h:
(threadSelfRestrictRWXToRW):
(threadSelfRestrictRWXToRX):
* Source/JavaScriptCore/jit/ExecutableAllocator.cpp:
(JSC::initializeJITPageReservation):
* Source/WTF/wtf/PlatformUse.h:
* Source/WebCore/Configurations/WebCore.xcconfig:

Canonical link: https://commits.webkit.org/273590@main
